File: filepicker/src/main/java/droidninja/filepicker/PickerManager.java
package droidninja.filepicker;

import java.util.ArrayList;

import droidninja.filepicker.models.BaseFile;
import droidninja.filepicker.models.FileType;
import droidninja.filepicker.models.sort.SortingTypes;
import droidninja.filepicker.utils.Orientation;
import java.util.ArrayList;
import java.util.LinkedHashSet;

/**
 * Created by droidNinja on 29/07/16.
 */
public class PickerManager {
  private static PickerManager ourInstance = new PickerManager();
  private int maxCount = FilePickerConst.DEFAULT_MAX_COUNT;
  private boolean showImages = true;
  private int cameraDrawable = R.drawable.ic_camera;
  private SortingTypes sortingType = SortingTypes.none;

  public static PickerManager getInstance() {
    return ourInstance;
  }

  private ArrayList<String> mediaFiles;
  private ArrayList<String> docFiles;

  private LinkedHashSet<FileType> fileTypes;

  private int theme = R.style.LibAppTheme;

  private String title = null;

  private boolean showVideos;

  private boolean showGif;

  private boolean showSelectAll = false;

  private boolean docSupport = true;

  private boolean enableCamera = true;

  private Orientation orientation = Orientation.UNSPECIFIED;

  private boolean showFolderView = true;

  private String providerAuthorities;

  private PickerManager() {
    mediaFiles = new ArrayList<>();
    docFiles = new ArrayList<>();
    fileTypes = new LinkedHashSet<>();
  }

  public void setMaxCount(int count) {
    reset();
    this.maxCount = count;
  }

  public int getMaxCount() {
    return maxCount;
  }

  public int getCurrentCount() {
    return mediaFiles.size() + docFiles.size();
  }

  public void add(String path, int type) {
    if (path != null && shouldAdd()) {
      if (!mediaFiles.contains(path) && type == FilePickerConst.FILE_TYPE_MEDIA) {
        mediaFiles.add(path);
      } else if (!docFiles.contains(path) && type == FilePickerConst.FILE_TYPE_DOCUMENT) {
        docFiles.add(path);
      } else {
        return;
      }
    }
  }

  public void add(ArrayList<String> paths, int type) {
    for (int index = 0; index < paths.size(); index++) {
      add(paths.get(index), type);
    }
  }

  public void remove(String path, int type) {
    if ((type == FilePickerConst.FILE_TYPE_MEDIA) && mediaFiles.contains(path)) {
      mediaFiles.remove(path);
    } else if (type == FilePickerConst.FILE_TYPE_DOCUMENT) {
      docFiles.remove(path);
    }
  }

  public boolean shouldAdd() {
    if (maxCount == -1) return true;
    return getCurrentCount() < maxCount;
  }

  public ArrayList<String> getSelectedPhotos() {
    return mediaFiles;
  }

  public ArrayList<String> getSelectedFiles() {
    return docFiles;
  }

  public ArrayList<String> getSelectedFilePaths(ArrayList<BaseFile> files) {
    ArrayList<String> paths = new ArrayList<>();
    for (int index = 0; index < files.size(); index++) {
      paths.add(files.get(index).getPath());
    }
    return paths;
  }

  public void reset() {
    docFiles.clear();
    mediaFiles.clear();
    fileTypes.clear();
    maxCount = -1;
  }

  public void clearSelections() {
    mediaFiles.clear();
    docFiles.clear();
  }

  public void deleteMedia(ArrayList<String> paths) {
    mediaFiles.removeAll(paths);
  }

  public int getTheme() {
    return theme;
  }

  public void setTheme(int theme) {
    this.theme = theme;
  }

  public String getTitle() {
    return title;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public boolean showVideo() {
    return showVideos;
  }

  public void setShowVideos(boolean showVideos) {
    this.showVideos = showVideos;
  }

  public boolean showImages() {
    return showImages;
  }

  public void setShowImages(boolean showImages) {
    this.showImages = showImages;
  }

  public boolean isShowGif() {
    return showGif;
  }

  public void setShowGif(boolean showGif) {
    this.showGif = showGif;
  }

  public boolean isShowFolderView() {
    return showFolderView;
  }

  public void setShowFolderView(boolean showFolderView) {
    this.showFolderView = showFolderView;
  }

  public void addFileType(FileType fileType) {
    fileTypes.add(fileType);
  }

  public void addDocTypes() {
    String[] pdfs = { "pdf" };
    fileTypes.add(new FileType(FilePickerConst.PDF, pdfs, R.drawable.icon_file_pdf));

    String[] docs = { "doc", "docx", "dot", "dotx" };
    fileTypes.add(new FileType(FilePickerConst.DOC, docs, R.drawable.icon_file_doc));

    String[] ppts = { "ppt", "pptx" };
    fileTypes.add(new FileType(FilePickerConst.PPT, ppts, R.drawable.icon_file_ppt));

    String[] xlss = { "xls", "xlsx" };
    fileTypes.add(new FileType(FilePickerConst.XLS, xlss, R.drawable.icon_file_xls));

    String[] txts = { "txt" };
    fileTypes.add(new FileType(FilePickerConst.TXT, txts, R.drawable.icon_file_unknown));
  }

  public ArrayList<FileType> getFileTypes() {
    return new ArrayList<>(fileTypes);
  }

  public boolean isDocSupport() {
    return docSupport;
  }

  public void setDocSupport(boolean docSupport) {
    this.docSupport = docSupport;
  }

  public boolean isEnableCamera() {
    return enableCamera;
  }

  public void setEnableCamera(boolean enableCamera) {
    this.enableCamera = enableCamera;
  }

  public Orientation getOrientation() {
    return orientation;
  }

  public void setOrientation(Orientation orientation) {
    this.orientation = orientation;
  }

  public String getProviderAuthorities() {
    return providerAuthorities;
  }

  public void setProviderAuthorities(String providerAuthorities) {
    this.providerAuthorities = providerAuthorities;
  }

  public void setCameraDrawable(int drawable) {
    this.cameraDrawable = drawable;
  }

  public int getCameraDrawable() {
    return cameraDrawable;
  }

  public boolean hasSelectAll() {
    return maxCount == -1 && showSelectAll;
  }

  public void enableSelectAll(boolean showSelectAll) {
    this.showSelectAll = showSelectAll;
  }

  public SortingTypes getSortingType() {
    return sortingType;
  }

  public void setSortingType(SortingTypes sortingType) {
    this.sortingType = sortingType;
  }
}

File: filepicker/src/main/java/droidninja/filepicker/PickerManager.kt
package droidninja.filepicker

import android.content.pm.ActivityInfo
import java.util.ArrayList

import droidninja.filepicker.models.BaseFile
import droidninja.filepicker.models.FileType
import droidninja.filepicker.models.sort.SortingTypes
import java.util.LinkedHashSet

/**
 * Created by droidNinja on 29/07/16.
 */
object PickerManager {
    private var maxCount = FilePickerConst.DEFAULT_MAX_COUNT
    private var showImages = true
    var cameraDrawable = R.drawable.ic_camera
    var sortingType = SortingTypes.none

    val selectedPhotos: ArrayList<String> = ArrayList()
    val selectedFiles: ArrayList<String> = ArrayList()

    private val fileTypes: LinkedHashSet<FileType> = LinkedHashSet()

    var theme : Int = R.style.LibAppTheme

    var title: String? = null

    private var showVideos: Boolean = false

    var isShowGif: Boolean = false

    private var showSelectAll = false

    var isDocSupport = true
        get() = field

    var isEnableCamera = true

    /**
     * The preferred screen orientation this activity would like to run in.
     * From the {@link android.R.attr#screenOrientation} attribute, one of
     * {@link #SCREEN_ORIENTATION_UNSPECIFIED},
     * {@link #SCREEN_ORIENTATION_LANDSCAPE},
     * {@link #SCREEN_ORIENTATION_PORTRAIT},
     * {@link #SCREEN_ORIENTATION_USER},
     * {@link #SCREEN_ORIENTATION_BEHIND},
     * {@link #SCREEN_ORIENTATION_SENSOR},
     * {@link #SCREEN_ORIENTATION_NOSENSOR},
     * {@link #SCREEN_ORIENTATION_SENSOR_LANDSCAPE},
     * {@link #SCREEN_ORIENTATION_SENSOR_PORTRAIT},
     * {@link #SCREEN_ORIENTATION_REVERSE_LANDSCAPE},
     * {@link #SCREEN_ORIENTATION_REVERSE_PORTRAIT},
     * {@link #SCREEN_ORIENTATION_FULL_SENSOR},
     * {@link #SCREEN_ORIENTATION_USER_LANDSCAPE},
     * {@link #SCREEN_ORIENTATION_USER_PORTRAIT},
     * {@link #SCREEN_ORIENTATION_FULL_USER},
     * {@link #SCREEN_ORIENTATION_LOCKED},
     */
    var orientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
        get() = field

    var isShowFolderView = true

    var providerAuthorities: String? = null

    val currentCount: Int
        get() = selectedPhotos.size + selectedFiles.size

    fun setMaxCount(count: Int) {
        reset()
        this.maxCount = count
    }

    fun getMaxCount(): Int {
        return maxCount
    }

    fun add(path: String?, type: Int) {
        if (path != null && shouldAdd()) {
            if (!selectedPhotos.contains(path) && type == FilePickerConst.FILE_TYPE_MEDIA) {
                selectedPhotos.add(path)
            } else if (!selectedFiles.contains(path) && type == FilePickerConst.FILE_TYPE_DOCUMENT) {
                selectedFiles.add(path)
            } else {
                return
            }
        }
    }

    fun add(paths: ArrayList<String>, type: Int) {
        for (index in paths.indices) {
            add(paths[index], type)
        }
    }

    fun remove(path: String, type: Int) {
        if (type == FilePickerConst.FILE_TYPE_MEDIA && selectedPhotos.contains(path)) {
            selectedPhotos.remove(path)
        } else if (type == FilePickerConst.FILE_TYPE_DOCUMENT) {
            selectedFiles.remove(path)
        }
    }

    fun shouldAdd(): Boolean {
        return if (maxCount == -1) true else currentCount < maxCount
    }

    fun getSelectedFilePaths(files: ArrayList<BaseFile>): ArrayList<String> {
        val paths = ArrayList<String>()
        for (index in files.indices) {
            paths.add(files[index].path)
        }
        return paths
    }

    fun reset() {
        selectedFiles.clear()
        selectedPhotos.clear()
        fileTypes.clear()
        maxCount = -1
    }

    fun clearSelections() {
        selectedPhotos.clear()
        selectedFiles.clear()
    }

    fun deleteMedia(paths: ArrayList<String>) {
        selectedPhotos.removeAll(paths)
    }

    fun showVideo(): Boolean {
        return showVideos
    }

    fun setShowVideos(showVideos: Boolean) {
        this.showVideos = showVideos
    }

    fun showImages(): Boolean {
        return showImages
    }

    fun setShowImages(showImages: Boolean) {
        this.showImages = showImages
    }

    fun addFileType(fileType: FileType) {
        fileTypes.add(fileType)
    }

    fun addDocTypes() {
        val pdfs = arrayOf("pdf")
        fileTypes.add(FileType(FilePickerConst.PDF, pdfs, R.drawable.icon_file_pdf))

        val docs = arrayOf("doc", "docx", "dot", "dotx")
        fileTypes.add(FileType(FilePickerConst.DOC, docs, R.drawable.icon_file_doc))

        val ppts = arrayOf("ppt", "pptx")
        fileTypes.add(FileType(FilePickerConst.PPT, ppts, R.drawable.icon_file_ppt))

        val xlss = arrayOf("xls", "xlsx")
        fileTypes.add(FileType(FilePickerConst.XLS, xlss, R.drawable.icon_file_xls))

        val txts = arrayOf("txt")
        fileTypes.add(FileType(FilePickerConst.TXT, txts, R.drawable.icon_file_unknown))
    }

    fun getFileTypes(): ArrayList<FileType> {
        return ArrayList(fileTypes)
    }

    fun hasSelectAll(): Boolean {
        return maxCount == -1 && showSelectAll
    }

    fun enableSelectAll(showSelectAll: Boolean) {
        this.showSelectAll = showSelectAll
    }
}
